nb2workflow/nbadapter.py

```diff
@@ -515,7 +521,9 @@ def _execute(self, parameters, progress_bar = True, log_output = True, inplace=F
                 continue   
 
             break
-
+        if token:
+            # cleanup
+            os.environ['TOKEN'] = prev_token
```

Comment:
Cleanup was moved to "finally" block. I have manually tested the cleanup by inserting the exception raising code inside try ... finally. As for the thread-safety I didn't explored it.  @dsavchenko Do you have any suggestions how to do this?

Reply:
within the async job just before running papermill run, maybe add to logging PID and ThreadID.

Reply:
@volodymyrss Indeed I have verified that since threading.Thread is used to start AsyncWorker, different workers share the same PID, so we need to invent some other mechanism to pass token. I suggest using files similar to callback passing. Perhaps we could generate context.json file which may contain both callback and token, and perhaps something else in future.